docs(design): add latency definition, records service and visualization

hsgwa opened the pull request:
Signed-off-by: hsgwa <19860128+hsgwa@users.noreply.github.com>

hsgwa reviewed (commented):
removed

takam5f2 commented on docs/design/latency_definitions/index.md:
```diff
@@ -0,0 +1,141 @@
+
+Here, each element indicates the following
+
+- UseCode is a callback
```
```suggestion
- UserCode is a callback
```

hsgwa replied:
fixed in :[b104708](https://github.com/tier4/CARET_doc/pull/106/commits/b104708b0378c76f80d2ba457892ff9ad0dbfdc4)

takam5f2 commented on docs/design/latency_definitions/index.md:
```diff
@@ -0,0 +1,141 @@
+A message is received in the subscription callback and the processed data is published to the next node.
+In this way, information is propagated from the sensor node to the actuator node.
+
+CARET measures the following three main locations for latency calculation.
```
```suggestion
CARET samples events for latency calculation. The following three items are main types of the events.
```

hsgwa replied:
addressed in: [b104708](https://github.com/tier4/CARET_doc/pull/106/commits/b104708b0378c76f80d2ba457892ff9ad0dbfdc4)

takam5f2 commented on docs/design/latency_definitions/index.md:
```diff
@@ -0,0 +1,141 @@
+- Callback end
+- Publish
+
+Latency is calculated by calculating the time difference required for each measurement section.
```
```suggestion
Difference of timestamp between two events are corresponded to latency.
```

hsgwa replied:
addressed in: [b104708](https://github.com/tier4/CARET_doc/pull/106/commits/b104708b0378c76f80d2ba457892ff9ad0dbfdc4)

takam5f2 reviewed (changes requested):
I think that I didn't give you all comments. I will add new comments while I continue reviewing this PR.

takam5f2 commented on docs/design/latency_definitions/index.md:
```diff
@@ -0,0 +1,141 @@
+- [Node](./node.md)
+- [Path](./path.md)
+
+CARET can also check other time series data on the Publish side as well.
```
@hsgwa  I somehow the structure of this section.

```suggestion
CARET is capable of delivering data set of latency whose format is time series.
```

hsgwa replied:
addressed in: [b104708](https://github.com/tier4/CARET_doc/pull/106/commits/b104708b0378c76f80d2ba457892ff9ad0dbfdc4)

takam5f2 commented on docs/design/latency_definitions/index.md:
```diff
@@ -0,0 +1,141 @@
+- [Path](./path.md)
+
+CARET can also check other time series data on the Publish side as well.
+All the objects for which time series data can be obtained are listed below.
```
```suggestion
All of the objects for which CARET retrieve latency are listed below
```

hsgwa replied:
addressed in: [b104708](https://github.com/tier4/CARET_doc/pull/106/commits/b104708b0378c76f80d2ba457892ff9ad0dbfdc4)

takam5f2 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,40 @@
+# Callback
+
+The callback latency is the time difference from callback start to callback end.
```
```suggestion
Callback latency is defined as duration between beginning and end of callback execution.
```

takam5f2 replied:
`Time difference` has meaning for expressing difference between two time zone. It is not appropriate word here.
https://www.collinsdictionary.com/jp/dictionary/english/time-difference

hsgwa replied:
addressed: [ea6f556](https://github.com/tier4/CARET_doc/pull/106/commits/ea6f556a84802265fdd6222bee05a63aa0620a68)

takam5f2 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,40 @@
+l_{\rm{callback}} = t_{\rm{callback\ end}} - t_{\rm{callback\ start}}
+$$
+
+A simplified sequence diagram focusing only on the relevant data flow is shown below.
```
I think that it is confusing the sequence diagram is designed to show data flow. 

```suggestion
The sequence diagram shows how CARET rclcpp pick up two events; callback_start and callback_end 
```

hsgwa replied:
addressed: [ea6f556](https://github.com/tier4/CARET_doc/pull/106/commits/ea6f556a84802265fdd6222bee05a63aa0620a68)

takam5f2 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,40 @@
+
+activate Rclcpp
+
+Rclcpp -> LTTng : callback start
```
sample `callback_start`

takam5f2 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,40 @@
+activate Callback
+Callback -> Rclcpp
+deactivate Callback
+Rclcpp -> LTTng : callback end
```
sample callback_end

takam5f2 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,40 @@
+
+activate Rclcpp
+
+Rclcpp -> LTTng : callback start
```
```suggestion
Rclcpp -> LTTng : sample callback_start
```

hsgwa replied:
addressed: [ea6f556](https://github.com/tier4/CARET_doc/pull/106/commits/ea6f556a84802265fdd6222bee05a63aa0620a68)

takam5f2 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,40 @@
+activate Callback
+Callback -> Rclcpp
+deactivate Callback
+Rclcpp -> LTTng : callback end
```
```suggestion
Rclcpp -> LTTng : sample callback_end
```

hsgwa replied:
addressed: [ea6f556](https://github.com/tier4/CARET_doc/pull/106/commits/ea6f556a84802265fdd6222bee05a63aa0620a68)

takam5f2 commented on docs/design/latency_definitions/communication.md:
```diff
@@ -0,0 +1,170 @@
+# Communication
+
+The communication latency is the time difference between the time the message is passed from the ownership of the callback to the next callback.
```
```suggestion
Communication latency is an expression of how much time it takes for a topic message to travel from source callback to next callback.
```

hsgwa replied:
28579f5

takam5f2 commented on docs/design/latency_definitions/communication.md:
```diff
@@ -0,0 +1,170 @@
+The communication latency is the time difference between the time the message is passed from the ownership of the callback to the next callback.
+
+$$
+l_{comm} = t_{receive} - t_{transfer}
```
```suggestion
l_{comm} = t_{reception} - t_{transmission}
```

hsgwa replied:
removed: 
28579f5

takam5f2 commented on docs/design/latency_definitions/communication.md:
```diff
@@ -0,0 +1,170 @@
+l_{comm} = t_{receive} - t_{transfer}
+$$
+
+In particular, if transfer of ownership is publish and receive of ownership is subscribe, it is defined as follows
```
As transmission is replaced by publish, but reception is subscription in the context of ROS 2, the expression is translated into the next one.

takam5f2 commented on docs/design/latency_definitions/communication.md:
```diff
@@ -0,0 +1,170 @@
+l_{comm} = t_{receive} - t_{transfer}
+$$
+
+In particular, if transfer of ownership is publish and receive of ownership is subscribe, it is defined as follows
```
```suggestion
As transmission is replaced by publish and reception is subscription in the context of ROS 2, the equation is translated into the next one.
```

hsgwa replied:
removed: 
28579f5

takam5f2 commented on docs/design/latency_definitions/communication.md:
```diff
@@ -0,0 +1,170 @@
+
+<prettier-ignore-start>
+!!! Info
+        In this definition, communication latency is now affected by scheduling.
```
```suggestion
        In this definition, communication latency is now affected by scheduling latency.
```

hsgwa replied:
28579f5

takam5f2 commented on docs/design/latency_definitions/communication.md:
```diff
@@ -0,0 +1,170 @@
+!!! Info
+        In this definition, communication latency is now affected by scheduling.
+        It is not only the communication latency of the DDS.
+        If multiple callbacks can be executed at the same time, it may include the execution time of other callbacks.
```
I cannot understand the intention of this sentence.

hsgwa replied:
removed 
28579f5

takam5f2 commented on docs/design/latency_definitions/path.md:
```diff
@@ -0,0 +1,24 @@
+# Path
+
+Path latency is the sum of node latency and communication latency included in the path.
```
```suggestion
Path latency is sum of node latency and communication latency included in a certain path.
```

takam5f2 commented on docs/design/.pages:
```diff
@@ -1,29 +1,17 @@
-            - Overview: design/configuration
-            - File format: design/configuration/architecture_file
+        - Design: configuration
+    - Latency definitions:
```
```suggestion
    - Event and latency definition:
```

takam5f2 replied:
@hsgwa  
Could you rename `latency_definitions` directory to `event_and_latency_definition`?

keita1523 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+The coarser granularity, the more time information is aggregated to evaluate the measurement as a whole.  
+The most granular statistics are suitable for regression testing.
+
+On the other hand, the finer granularity, the more detailed information is expressed, so it is more suitable for analysis. evaluation
+Latency and response time are some of the indicators to evaluate performance.  
```
```
for analysis. evaluation
```
Typo.

takam5f2 replied:
```suggestion
Latency and response time are mainly used metrics to evaluate performance.  
```

keita1523 commented on docs/design/latency_definitions/index.md:
```diff
@@ -0,0 +1,141 @@
+| Target                              | Configuration required? |
+| ----------------------------------- | ----------------------- |
+| [Path](./path.md)                   | Yes                     |
+| [NodePath](./node.md)               | Yes                     |
```
Why NodePath instead of Node?

keita1523 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,40 @@
+The callback latency is the time difference from callback start to callback end.
+
+$$
+l_{\rm{callback}} = t_{\rm{callback\ end}} - t_{\rm{callback\ start}}
```
reloaded, then showed up correctly.

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+# Node
+
+Node latency is the time difference between when a callback takes ownership and when it abandons.
```
```suggestion
Node latency is elapsed time from message subscription to message publish.
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+l_{node} = t_{transfer} - t_{receive}
+$$
+
+In particular, if transfer of ownership is subscribe and receive of ownership is publish, it is defined as follows
```
```suggestion
As transmission is replaced by publish and reception is subscription in the context of ROS 2, the equation is translated into the next one.
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+l_{node} = t_{pub} - t_{sub}
+$$
+
+See [Communication](./communication.md) for the reason ownership is used to define latency.
```
Could you remove this sentence?

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+
+![message context](../../imgs/message_context.drawio.png)
+
+Here, the time difference between Subscribe and Publish can be calculated as node latency.  
```
```suggestion
Here, the elapsed time from subscription is defined as node latency.  
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+
+Here, the time difference between Subscribe and Publish can be calculated as node latency.  
+In this way, node latency can be calculated when message dependencies are defined.
+6
```
```suggestion

```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+In this way, node latency can be calculated when message dependencies are defined.
+6
+In the previous example, we presented a case where a node is subscribed and immediately published.
+In actuality, the Message context can be quite complex, because node processing can be defined by the developer.
```
```suggestion
Actually, the message context might be quite complex when dependency between the incoming message and the outgoing one is complicated on multiple callbacks.
```

takam5f2 replied:
```suggestion
Actually, the message context can be quite complex because data processing in a node is freely defined by an application developer.
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+
+![complex message context](../../imgs/message_context_complex.drawio.png)
+
+Each case is described below.
```
```suggestion
Examples of complex message context are described below.
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+
+Each case is described below.
+
+- Buffering, for example, is a buffer delay process.
+- Multiple use is for moving average processing, for example. There are multiple candidates for node latency.
+- Unused is a message that has not been published and has not been used. This is evaluated as a kind of message drop.
+- Cross can occur in cases where messages are retrieved based on the timestamp of the message rather than the system time.
```
How about renaming as blow to align participle of words;

- Buffering -> Buffered
- Multiple use -> Multiple inputted
- Unused -> Unused (not changed)
- cross -> Crossed

takam5f2 replied:
Could you change the corresponding figure?

hsgwa replied:
addressed  [58b88f0](https://github.com/tier4/CARET_doc/pull/106/commits/58b88f078e12f475c24b0efda0feb1b235e99772)

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+- Unused is a message that has not been published and has not been used. This is evaluated as a kind of message drop.
+- Cross can occur in cases where messages are retrieved based on the timestamp of the message rather than the system time.
+
+In either case, it is difficult to automatically determine the message context.
```
```suggestion
In any of those cases, it is difficult to automatically determine the message context.
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+
+In either case, it is difficult to automatically determine the message context.
+
+In some cases this can be covered by Configuration.
```
```suggestion
the message context are given for some cases of them as shown in configuration.
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+In some cases this can be covered by Configuration.
+For more information, see [Configuration](. /... /configuration/index.md) for details.
+
+If Configuration does not cover the situation, we are considering a mechanism to allow users to describe message dependencies as a future plan.
```
```suggestion
To enlarge coverage the others of complex cases, we are considering a mechanism to allow users to describe message dependencies in their user codes.
```

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,58 @@
+For more information, see [Configuration](. /... /configuration/index.md) for details.
+
+If Configuration does not cover the situation, we are considering a mechanism to allow users to describe message dependencies as a future plan.
+For details, please refer to [TILDE](. /software_architecture/tilde.md) for details.
```
```suggestion
[TILDE](. /software_architecture/tilde.md) is one of good candidates to achieve it.
```

takam5f2 commented on docs/design/latency_definitions/path.md:
```diff
@@ -0,0 +1,24 @@
+l_{comm} = t_{sub} - t_{pub} \\
+$$
+
+Here, path is one of the paths chosen from the node graph.
```
Could you tell me how `the node graph` is related to `the paths`?

takam5f2 replied:
Thanks, but I'd like to improve this sentence at next comment.

takam5f2 commented on docs/design/latency_definitions/path.md:
```diff
@@ -0,0 +1,24 @@
+
+<prettier-ignore-start>
+!!! Info
+    In the above definition, the start time of the latency is the start node's Publish and the end time of the latency measurement is the end node's Subscribe.
```
```suggestion
    In the above definition, the starting time of the latency is publish of the starting node and the end time of the latency is subscription of the end node.
```

takam5f2 commented on docs/design/latency_definitions/publisher.md:
```diff
@@ -0,0 +1,79 @@
+
+The Publisher provides the time and relevant information when publishing.
+
+A simplified sequence diagram focusing only on the relevant data flow is shown below.
```
```suggestion
A simplified sequence diagram focusing only on the relevant events is shown below.
```

takam5f2 commented on docs/design/latency_definitions/subscription.md:
```diff
@@ -0,0 +1,94 @@
+# Subscription
+
```
```suggestion

Topic message is received by subscription via inter-process communication or intra-process one.
Then, `Subscription` object has common data fields, but different value is filled in `source_timestamp` field.


```

takam5f2 commented on docs/design/latency_definitions/subscription.md:
```diff
@@ -0,0 +1,94 @@
+
+## Inter process communication
+
+A simplified sequence diagram focusing only on the relevant data flow is shown below.
```
```suggestion
A simplified sequence diagram focusing only on the relevant events is shown below.
```

takam5f2 commented on docs/design/latency_definitions/timer.md:
```diff
@@ -0,0 +1,59 @@
+
+Timer provides the time at which the Timer callback timeout and related information.
+
+A simplified sequence diagram focusing only on the relevant data flow is shown below.
```
```suggestion
A simplified sequence diagram focusing only on the relevant events is shown below.
```

takam5f2 commented on docs/design/processing_trace_data/records.md:
```diff
@@ -1,11 +1,20 @@
+| ...                      | ...                    |
 
-This sections describes Record object which stores trace data and latency is.
+This table is processed to calculate latency, period, etc (See [Records Service](./records_service.md)).
```
```suggestion
This table is referred to calculate latency, period, and etc. (See [Records Service](./records_service.md)).
```

takam5f2 replied:
@hsgwa  
I think that latency is calculated from this table. If this is true, `refer` might be more appropriate word.

takam5f2 commented on docs/design/processing_trace_data/records.md:
```diff
@@ -1,11 +1,20 @@
 
-In CARET, trace data is stored in tables, and latency calculation is performed by processing the tables.
-In addition to general table join processing, CARET defines classes which has originally defined join processing for latency calculation.
+CARET stores trace data as a table as shown below.
```
```suggestion
CARET provides trace data to user. The most primitive format is a table per event, which is picked up by tracepoint, shown as below.
```

takam5f2 commented on docs/design/processing_trace_data/records.md:
```diff
@@ -1,11 +1,20 @@
-This sections describes Record object which stores trace data and latency is.
+This table is processed to calculate latency, period, etc (See [Records Service](./records_service.md)).
+
+To create the above table, merging processes are performed with the traced results.
```
```suggestion
The most primitive format is a table per event which picked up by a corresponding tracepoint. Merging multiple event tables makes a new table for metrics.
```

takam5f2 commented on docs/design/processing_trace_data/records.md:
```diff
@@ -1,11 +1,20 @@
+This table is processed to calculate latency, period, etc (See [Records Service](./records_service.md)).
+
+To create the above table, merging processes are performed with the traced results.
+In addition to general table join processing, CARET defines classes which has originally defined join processing for latency calculation.
```
```suggestion
In addition to simple table merging, CARET defines classes which has originally defined merging method for latency calculation.
```

takam5f2 replied:
I think that there is no difference; merging and joining, isn't it?

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+# Visualizations
+
+CARET is a tool for performance evaluation.
```
```suggestion
CARET is a tool for performance evaluation and analysis.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+For evaluation, quantification and visualization would be sufficient.  
+For analysis, on the other hand, various factors are involved in behavior and performance, so it is necessary to investigate from multiple perspectives.
+
+CARET has multiple visualizations for evaluation and analysis APIs that are readily available.  
```
```suggestion
CARET has multiple visualizations for evaluation and analysis APIs.  
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+For analysis, on the other hand, various factors are involved in behavior and performance, so it is necessary to investigate from multiple perspectives.
+
+CARET has multiple visualizations for evaluation and analysis APIs that are readily available.  
+Also, it's also possible to directly acquire and evaluate the data on your own (see [Processing trace data](../processing_trace_data/index.md) for detail).
```
```suggestion
Also, CARET let you to directly acquire and evaluate the data on your own (see [Processing trace data](../processing_trace_data/index.md) for detail).
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+
+### Evaluation flow
+
+CARET has a large amount of information that can be obtained, and it is difficult to evaluate large-scale applications simply by displaying the recorded trace data in time series.
```
```suggestion
CARET has a large amount of recorded events, which prevents you from investigating performance of your application smoothly. Just displaying recorded events thoughtlessly might waste your time.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+![Analysis flow](../../imgs/analysis_flow.drawio.png)
+
+Here, the horizontal axis represents the visualization granularity and the vertical axis represents the processing granularity.  
+CARET leads from issue detection to cause identification by changing the processing and visualization granularity from the rough granularity in the upper right to the fine granularity in the lower left.
```
```suggestion
CARET help you to find issues and their causes by changing the processing and visualization granularity, respectively. You can approach an issues and its cause step by step from rough to fine granularity, from right-upper to left-lower in the figure.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+
+![Analysis flow](../../imgs/analysis_flow.drawio.png)
+
+Here, the horizontal axis represents the visualization granularity and the vertical axis represents the processing granularity.  
```
Please add the meaning of visualization granularity and the processing granularity.

takam5f2 replied:
I think that adding the following sentence might be kind.
```
Those definitions will be explained subsequent sections.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+The coarser granularity, the more time information is aggregated to evaluate the measurement as a whole.  
+The most granular statistics are suitable for regression testing.
+
+On the other hand, the finer granularity, the more detailed information is expressed, so it is more suitable for analysis. evaluation
```
```suggestion
On the other hand, the finer granularity, the more detailed information is expressed, so it is more suitable for analysis. 
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+
+![Processing Granularity](../../imgs/processing_granularity.drawio.png)
+
+The processing granularity can be ordered in coarse order as follows.
```
```suggestion
The processing granularity means granularity of subsystem in an application, it is ordered in coarse order as follows.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+- callbackCurrent CARET does not support measurement of arbitrary functions or DDS enqueue/dequeue.
+- function
+
+Path is suitable for evaluation of system performance, while node and callback are suitable for evaluation of functional performance.
```
```suggestion
Path is suitable for performance evaluation per system , while node and callback are suitable for  performance evaluation per component or smaller subsystem.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+
+<prettier-ignore-start>
+!!! Notice
+        Currently CARET does not support measurement of arbitrary functions or DDS enqueue/dequeue or system calls.
```
```suggestion
         In the present, CARET does not support measurement of arbitrary functions or DDS enqueue/dequeue or system calls.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+
+### Time domain and Frequency domain
+
+We described behavior and indicator in terms of visualization granularity, but there are multiple indicators, such as latency and frequency.  
```
```suggestion
We described behavior and indicator in terms of visualization granularity, but there are multiple indicators such as latency and frequency.  
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+### Time domain and Frequency domain
+
+We described behavior and indicator in terms of visualization granularity, but there are multiple indicators, such as latency and frequency.  
+When considering which indicator to use, it is useful to consider the time domain indicators and frequency domain indicators.
```
```suggestion
Metrics for performance analysis is categorized into two types; time-related metrics and frequency-related metrics. CARET recommends you to be aware two types.
```

takam5f2 commented on docs/design/visualizations/index.md:
```diff
@@ -0,0 +1,92 @@
+| Pros      | Easy to compare with system requirements | No need to define latency or path             |
+| Cons      | Need to define latency or path           | Difficult to compare with system requirements |
+
+In the table above, period is an indicator that expresses the time interval between each time, so it is classified in the frequency domain.
```
```suggestion
In the table above, period is metrics that expresses the time interval between sequential cyclic events, so it is classified into the frequency-related metrics.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+The Records object holds time-series data such as message flow in a table as shown below.
+
+| Start timestamp | ... | End timestamp |
```
```suggestion
| Initial timestamp | ... | Last timestamp |
```

takam5f2 replied:
I think that `start timestamp` and `end timestamp` are better because large character does not have meaning here.

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+| 3.0             | ... | 2.1           |
+| ...             | ... | ...           |
+
+The start_timestamp column contains the system time at the starting point.
```
```suggestion
The initial timestamp column contains the system time at the starting point.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+| ...             | ... | ...           |
+
+The start_timestamp column contains the system time at the starting point.
+The end_timestamp column contains the system time at the end point.
```
```suggestion
The last timestamp column contains the system time at the end point.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+The end_timestamp column contains the system time at the end point.
+Both columns indicate the system time, and if there is no corresponding value for start, the value is NaN.
+
+For the intermediate columns, the time at the intermediate point between start and end is represented.
```
```suggestion
For the intermediate columns, the time at the intermediate point between starting point and end one is represented.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+The rows of the table are visualized as lines in a message flow diagram.
+
+This section describes the process o
+The class that reads TraceData is the Records object.
```
Sorry, but I cannot understand what you meant.

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+- [Records](./records.md)
+
+The following are processing using Records object.
```
```suggestion
The followings are processing using Records object.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+## Period
+
+The period is defined as the time difference between the columns of interest.
```
```suggestion
Period is metrics defined as the elapsed time between two cyclic event. On the table above, subtraction a previous timestamp from a target timestamp on the same row means period.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+## Latency
+
+Latency is defined as the time difference between the two columns of interest.
```
```suggestion
Latency is is length of time from preceding event to afterward one. It is defined as difference from preceding timestamp to afterward on the same row.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+Response time is the amount of time it takes for a system to respond to an input.
+
+As shown above, latency can be calculated if a table can be constructed.
```
```suggestion
As shown above, latency can be calculated if a table is constructed.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+## Response Time
+
+Response time is the amount of time it takes for a system to respond to an input.
```
```suggestion
Response time is amount of time it takes for a system to respond to an input.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+Response time is the amount of time it takes for a system to respond to an input.
+
+As shown above, latency can be calculated if a table can be constructed.
+However, the calculated latency is not suitable for evaluating response time.
```
```suggestion
Latency is defined as delay from input to any output. This calculation is very simple, but it is not used for evaluating response. if multiple output depends on a single input, latency is not regarded as response time. 
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+As shown above, latency can be calculated if a table can be constructed.
+However, the calculated latency is not suitable for evaluating response time.
+For example, if a sensor is driven at 10 Hz, a latency of up to 100 ms should be considered.
```
```suggestion
Moreover, for example, if a sensor is driven at 10 Hz, a latency of up to 100 ms should be considered.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+However, the calculated latency is not suitable for evaluating response time.
+For example, if a sensor is driven at 10 Hz, a latency of up to 100 ms should be considered.
+
+CARET defines latency as the best-case response time and worst-case response time as the value that includes the delay due to cycles, as a more comprehensive approach.
```
```suggestion
CARET serves best-case response time and worst-case response time. The former is equal to latency from an input to a corresponding initial output. CARET give the latter one because CARET takes cycle latency into account.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+| 3.0             | 0.2 (3.2 - 3.0)         | 2.3 (3.2 - 1.0)          |
+| ...             | ...                     | ...                      |
+
+Note that Best-case response time has the same definition as Latency.
```
```suggestion
Note that Best-case response time is equal to Latency.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+![Latency vs Response time](../../imgs/latency_vs_response_time.drawio.png)
+
+Cases like the one shown above include
```
Of the pseudo two message flow diagram as shown above, the upper figure shows latency and the lower explains response time. 
The upper figure includes cumbersome cases as follows;

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+
+Cases like the one shown above include
+
+- Cases where multiple latencies are defined for a single output
```
Single output from multiple inputs

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+| 5.5             | 6.0           |
+| 5.5             | 6.5           |
+
+After leaving only the best-case flows as shown in the figure below,
```
```suggestion
After extracting only the best-case flows as shown in the lower figure,
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+| 5.5             | 6.5           |
+
+After leaving only the best-case flows as shown in the figure below,
+the best-case is calculated as the latency of the line and the worst-case as the latency calculated from the previous input.
```
```suggestion
the best case is calculated as the latency of the flow and the worst-case as the latency is regarded as delay from the previous input to the output.
```

takam5f2 commented on docs/design/processing_trace_data/records_service.md:
```diff
@@ -0,0 +1,214 @@
+After leaving only the best-case flows as shown in the figure below,
+the best-case is calculated as the latency of the line and the worst-case as the latency calculated from the previous input.
+
+The response time after processing is shown below.
```
```suggestion
According to those definitions, the response time is calculated as below.
```

takam5f2 commented on docs/design/latency_definitions/index.md:
```diff
@@ -0,0 +1,142 @@
+
+CARET provides time series data of events through Python objects.
+Time series data can be retrieved with the Python objects which have to_dataframe API.
+All objects capable of retrieving time-series data are listed below.
```
I'm sorry, but I took a simple mistake.

```suggestion
All objects are capable of retrieving time-series data are listed below.
```

hsgwa replied:
addressed

takam5f2 commented on docs/design/latency_definitions/callback.md:
```diff
@@ -0,0 +1,41 @@
+l_{\rm{callback}} = t_{\rm{callback\ end}} - t_{\rm{callback\ start}}
+$$
+
+The sequence diagram shows how CARET rclcpp pick up two events; callback_start and callback_end.
```
I'm sorry, but I took a simple mistake.

```suggestion
The sequence diagram shows how CARET rclcpp picks up two events; callback_start and callback_end.
```

hsgwa replied:
addressed

takam5f2 commented on docs/design/latency_definitions/callback.md:
````diff
@@ -0,0 +1,41 @@
+Rclcpp -> LTTng : sample callback_end
+```
+
+`to_dataframe` API returns following columns.
````
```suggestion
`to_dataframe` API returns a table which has the following columns.
```

hsgwa replied:
addressed

hsgwa replied:
addressed

takam5f2 commented on docs/design/latency_definitions/communication.md:
````diff
@@ -0,0 +1,155 @@
+@enduml
+```
+
+`to_dataframe` API returns following columns.
````
```suggestion
`to_dataframe` API returns a table which has the following columns.
```

hsgwa replied:
addressed

takam5f2 commented on docs/design/latency_definitions/node.md:
```diff
@@ -0,0 +1,56 @@
+
+![message context](../../imgs/message_context.drawio.png)
+
+Here, the elapsed time from subscription is defined as node latency.  
```
Sorry, but I had taken a simple mistake.
```suggestion
Here, the elapsed time from subscription to publish is defined as node latency.  
```

takam5f2 commented on docs/design/latency_definitions/path.md:
```diff
@@ -0,0 +1,24 @@
+
+<prettier-ignore-start>
+!!! Info
+    In the above definition, the starting time of the latency is publish of the initial node and the end time of the latency is subscription of the last node.
```
Sorry, but I took a mistake.

```suggestion
    In the above definition, the star time of the latency is publish of the initial node and the end time of the latency is subscription of the last node.
```

takam5f2 commented on docs/design/latency_definitions/publisher.md:
```diff
@@ -0,0 +1,79 @@
+# Publisher
+
+`Publisher`-based object provides a set of timestamps when publishing.
```
```suggestion
`Publisher`-based object provides a set of timestamps collected during message publish.
```

takam5f2 commented on docs/design/latency_definitions/publisher.md:
````diff
@@ -0,0 +1,79 @@
+@enduml
+```
+
+`to_dataframe` API returns following columns.
````
```suggestion
`to_dataframe` API returns a table which has the following columns.
```

takam5f2 commented on docs/design/latency_definitions/subscription.md:
````diff
@@ -0,0 +1,97 @@
+@enduml
+```
+
+`to_dataframe` API returns following columns.
````
```suggestion
`to_dataframe` API returns a table which has the following columns.
```

takam5f2 commented on docs/design/latency_definitions/communication.md:
````diff
@@ -0,0 +1,155 @@
+@enduml
+```
+
+`to_dataframe` API returns following columns.
````
```suggestion
`to_dataframe` API returns a table which has the following columns.
```

takam5f2 commented on docs/design/latency_definitions/subscription.md:
````diff
@@ -0,0 +1,97 @@
+@enduml
+```
+
+`to_dataframe` API returns following columns.
````
```suggestion
`to_dataframe` API returns a table which has the following columns.
```

takam5f2 commented on docs/design/latency_definitions/timer.md:
````diff
@@ -0,0 +1,59 @@
+@enduml
+```
+
+`to_dataframe` API returns following columns.
````
```suggestion
`to_dataframe` API returns a table which has the following columns.
```

takam5f2 commented on docs/design/latency_definitions/path.md:
```diff
@@ -0,0 +1,24 @@
+l_{comm} = t_{sub} - t_{pub} \\
+$$
+
+Here, path is one of the paths chosen from the node graph.
```
```suggestion
```

takam5f2 replied:
Moved https://github.com/tier4/CARET_doc/pull/106/files#r1031346098

hsgwa replied:
[2470a77](https://github.com/tier4/CARET_doc/pull/106/commits/2470a77b1216671bd87d9ba48d765b935799ac11)

takam5f2 commented on docs/design/latency_definitions/path.md:
```diff
@@ -0,0 +1,24 @@
+# Path
+
+Path latency is sum of node latency and communication latency included in a certain path.
```
```suggestion
Path latency is sum of node latency and communication latency included in a certain path. Path represents data flow which  is constructed of multiple nodes connected each other. 
```

hsgwa replied:
[2470a77](https://github.com/tier4/CARET_doc/pull/106/commits/2470a77b1216671bd87d9ba48d765b935799ac11)

takam5f2 commented:
@hsgwa  Thank you for checking my feedback.
I left some small comments. Please check them.

takam5f2 commented:
@keita1523  
Please resolve conversation if @hsgwa corresponded your comments as you expected.

hsgwa commented:
@takam5f2 
- adopted your suggestion.
- changed image [58b88f0](https://github.com/tier4/CARET_doc/pull/106/commits/58b88f078e12f475c24b0efda0feb1b235e99772)
- renamed directory [17d1289](https://github.com/tier4/CARET_doc/pull/106/commits/17d12897ac03519b77128dc1e9b7a382992fae9d)

takam5f2 reviewed (approved):
LGTM

takam5f2 commented:
@keita1523  
Could you approve this PR if you are satisfied with the PR?

keita1523 reviewed (approved):
LGTM.
